### PR TITLE
fix(web): dynamic font downscaling for OSK keys

### PR DESCRIPTION
**Pull request by jahorton**
From https://github.com/keymanapp/keyman/pull/4255#pullrequestreview-568783714:
> LGTM -- thank you for the additional screenshots; they all look better. I guess the next step would be to reduce font size to fit on the key but that is more complicated...

Yes... but it turns out we were already using some of the necessary components anyway.  A few tweaks and a lot of experimentation later, and we can have something even nicer than before!  I also cleaned up some things from the base PR, as the original approach had some issues once I started manipulating the font sizes.

Note that this uses some [pretty modern Web tech (MDN)](https://developer.mozilla.org/en-US/docs/Web/API/TextMetrics/fontBoundingBoxAscent#browser_compatibility), so it's important to have null guards and if-checks against the fancy `TextMetrics` properties that hadn't been previously used.  I did consider use of `actualBoundingBoxAscent` over `fontBoundingBoxAscent`, but there were far more "rough edges" and clipping scenarios this way.

The following comparison categories are as follows:
- "Before":  the "after" state from #4255.  
- "After (legacy)":  the result of this PR's changes when `fontBoundingBoxAscent` is not available.
- "After" / "After (modern)": the result of this PR's changes when `fontBoundingBoxAscent` _is_ available.

I'll be highlighting the most interesting cases here.

-----

### English (`us-1.0`, from KMW's test pages)

**Before**

Screen Shot 2021-01-14 at 3 52 45 PM

Screen Shot 2021-01-14 at 4 04 27 PM

**After (legacy)**

Screen Shot 2021-01-15 at 2 29 08 PM

Without font scaling data available... it's about the same as "before".

**After (modern)**

Screen Shot 2021-01-15 at 2 01 02 PM

Screen Shot 2021-01-15 at 2 01 20 PM

There is a bit of roughness at extreme scales; the bottom of descending characters does end up getting cut off if the font starts off excessively large in comparison to the available resolution.

### `sil_cameroon_qwerty`

**Before**

Screen Shot 2021-01-14 at 4 11 25 PM

**After (legacy)**

Screen Shot 2021-01-15 at 2 32 12 PM

**After (modern)**

Screen Shot 2021-01-15 at 2 02 41 PM

The one casualty I've seen from these changes - the fancy custom layer key uses its own CSS styling, which... isn't as compatible with these changes as I'd expect.  Even then... I think it's still an improvement over the "before" from #4255.

### Khmer (`khmer_angkor`)

**Before**

Screen Shot 2021-01-15 at 8 55 05 AM

Screen Shot 2021-01-15 at 9 32 16 AM

**After**

Screen Shot 2021-01-15 at 2 06 10 PM

Screen Shot 2021-01-15 at 2 06 45 PM

### Burmese (`basic_kbdmyan`)

**Before**

Screen Shot 2021-01-15 at 9 37 31 AM

**After**

Screen Shot 2021-01-15 at 2 08 39 PM

_Still_ not perfect... but it is quite close.  Again, some of the longest descenders are being cut off a bit.

### Devanagari (`basic_kbdindev`)

**Before**

Screen Shot 2021-01-15 at 9 39 21 AM

**After**

Screen Shot 2021-01-15 at 2 13 05 PM

### Cree, Plains - Syllabic (`nrc_crk_cans`)

**Before**

Screen Shot 2021-01-15 at 9 39 48 AM

**After**

Screen Shot 2021-01-15 at 2 14 39 PM

### Tibetan (basic_kbdtiprd)

**Before**

Screen Shot 2021-01-15 at 9 43 16 AM

**After**

Screen Shot 2021-01-15 at 2 15 54 PM

------------

Some extras for good measure:

- Key previews are _also_ scaled.  This one was a bit tricky and is sometimes imperfect.

Screen Shot 2021-01-15 at 2 21 08 PM

Screen Shot 2021-01-15 at 2 22 37 PM

Yeah... mostly b/c of descenders again.  At least it's pretty clearly within the right ballpark now.

- Bifocals not included.

Screen Shot 2021-01-15 at 2 24 44 PM

If we were actually worried about ultra-low resolution devices, I'd worry about scaling the 'underlying key' label's text.  (It's part of why that image is hard to read; the label's almost as big as each key's main text!)  But... since we're just using this approach to approximate the effect of large-scaled fonts... meh.

### CODE REVIEW

**Comment by jahorton** [comment]
So, when I say "pretty modern"... MDN says that Chrome for Android only supports the "modern" path starting with version 87.  That's the _latest stable version_, with 88 in beta as of a couple of days ago.

On the plus side, Apple's been _way_ ahead of the curve here.  Apparently iOS Safari has always supported it!

**Comment by mcdurdin** [comment]
The prior change (#4255) was good because it shifted characters into the key rect fairly seamlessly and should work across multiple browsers fairly consistently, and did not result in other side-effects.

I'm not comfortable with this change for beta, for a few reasons:

1. The dynamic scaling results in some pretty small text and I am guessing we'll see some keyboards that have fonts that break with this -- because of weird box sizes etc. These may be hard for us to find.
2. I am unhappy with different versions of Keyman and different browser versions for Keyman 14 producing such different results for the display -- this will be a support headache.
3. The change is not backward compatible so it's pretty hard to work around if you do have a font with funny box sizes that you need to use.

Given this, I think this should be a change we consider for a future version rather than at this stage in beta.

**Comment by jahorton** [comment]
>     1. The dynamic scaling results in some pretty small text and I am guessing we'll see some keyboards that have fonts that break with this -- because of weird box sizes etc. These may be hard for us to find.

To be fair... if the text is scaled because of this, it's near-certainly because the text was already "broken" from a user perspective.

...Unless the font's box info has issues that give bad measurements to the `TextMetrics` subsystem, true.

Even then, please remember that I'm using this as an analogue for scenarios with extremely large font sizes on... more "standard" resolutions.  If someone were _actually_ trying to use Keyman on a 200x150 device, it'd either be too small for OSK-based typing or the DPI would be crazy low, with each pixel taking up an unusually large amount of screen space compared to most monitors.

Which images contain uncomfortably "pretty small text", in your opinion?  To my eyes, anywhere with "small" text... it turns out to be quite necessary because of vertical space requirements of the font's characters.  I'll point to Burmese (`basic_kbdmyan`) as my primary example of such a case.  The least reasonable-looking one of these examples is probably the English (`us-1.0`) case, in my opinion.

>     2. I am unhappy with different versions of Keyman and different browser versions for Keyman 14 producing such different results for the display -- this will be a support headache.

I... uh... what?  Are we saying that you're not comfortable with this being a Web/Android/iOS only thing?  Otherwise... I'm not quite sure what you're referring to here.  Please elaborate.

File: web/source/osk/visualKeyboard.ts [hunk 1]
```diff
@@ -1805,11 +1834,19 @@ namespace com.keyman.osk {
       } else {
         let emSizeStr = getComputedStyle(document.body).fontSize;
         let emSize = util.getFontSizeStyle(emSizeStr).val;
+
         var emScale = 1;
         if(!this.isStatic) {
           // Reading this requires the OSK to be active, so we filter out
           // BuildVisualKeyboard calls here.
-          emScale = util.getFontSizeStyle(keyman.osk._Box).val;
+          let boxFontStyle = util.getFontSizeStyle(keyman.osk._Box);
```
Comment (mcdurdin) [inline web/source/osk/visualKeyboard.ts:1842]: So the only additional change I'd like to see is to `util.getFontSizeStyle`:

```js
     getFontSizeStyle(e: HTMLElement|string): {val: number, absolute: boolean} {
       var val: number;
       var fs: string;
 
       if(typeof e == 'string') {
         fs = e;
       } else {
-        fs = e.style.fontSize;
+        fs = getComputedStyle(e).fontSize;
       }
...
```

That copes with the situation where no style is set on `_Box` (or other elements which `getFontSizeStyle` may be called for)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 14.0.227-beta](https://keyman.com/downloads/releases/beta/14.0.227)

**Comment by mcdurdin** [comment]
I updated the bulk renderer to allow me to review these. Images on left are before the fix; those on the right are after. This is running in the Chrome mobile emulator, iPhone 5. I have not noted all differences; some are minor and I am not worried about them, and I have only included one or two instances per script.

After submitting this report, I am working on:

1. Cleaning up the bulk renderer changes to submit as a PR (including a bug fix for KeymanWeb font/stylesheet management that I encountered in testing)
2. Testing a theory around font scale which may resolve the bulk of these issues (fingers crossed?)

The following keyboards have worse results with the new rendering. I have highlighted some of problems but not all.

* vm_tamil_typewriter, vm_tamil_modular (Similar results for other Tamil keyboards, as would be expected):
image

* sil_yupik_cyrillic:
image

* sil_tai_dam_lao:
image

* sil_cherokee_nation:
image

* rac_western_punjabi. This one is not particularly problematic -- but the characters are now *less* vertically centered than previously (true for a number of Arabic script keyboards):
image

* phoneticmalayalam. Malayalam keyboards were slightly uneven previously but it is exacerbated now:
image

* mozhi_malayalam. This one is particularly bad!
image

* malar_malayalam_inscript also:
image

* kannada_basic:
image

-------

* Secondary issue -- I note that the base key cap is also impacted by the font scale, which shouldn't be happening.
image

-------

Other notes (may need to be turned into issues against these keyboards):
* sil_tai_dam, Tai Dam (SIL) "Tai Heritage Pro" font does not appear to load the font.
image

* warang_citi has CSS that doesn't work well on iPhone:
image

**Comment by jahorton** [comment]
Actually... that secondary issue appears to be super-related to the primary issue.  The key cap is being scaled _only_ for keys that required scaling due to overly wide text in those screenshots... so there's a strong chance there's an accidental double-scaling causing the font-size issues.  Now to find where it's coming from...

**Comment by jahorton** [comment]
> * rac_western_punjabi. This one is not particularly problematic -- but the characters are now _less_ vertically centered than previously (true for a number of Arabic script keyboards):
>   image
> 
> *

I noticed this with another testing keyboard or two as well.  The issue is that there usually _are_ characters somewhere on the keyboard that do use the excess headspace... though sometimes, it's only there for diacritics once applied.

**Comment by jahorton** [comment]
>     * sil_cherokee_nation:
>       image

Would I be safe in assuming this was targeting an iOS tablet form factor?  I can only repro it locally when I select 'iPad' or 'iPad Pro' in Chrome emulation; everything looks fine for iOS phone-based form factors.

**Comment by jahorton** [comment]
And... yep, the surface-layer cause is what I expected:

Screen Shot 2021-01-22 at 8 58 06 AM

Just gotta prevent the `kmw-key` application of `font-size`.  `kmw-key-text` is the intended target.

**Comment by jahorton** [comment]
> * mozhi_malayalam. This one is particularly bad!
>   image

Of interest - part of why this one's particularly bad is because of how the key-text is specified.  Note the difference between the 'default' and 'shift' layers:

Screen Shot 2021-01-22 at 9 33 21 AM

Screen Shot 2021-01-22 at 9 33 30 AM

The difference?  Note how the 'Q' key is defined on the 'default' layer:

Screen Shot 2021-01-22 at 9 34 44 AM

Naturally, HTML being HTML, the excess whitespace seen within the 'span' doesn't directly affect the key's appearance and alignment.  The problem, though - it _does_ affect the `TextMetric` return values.  The excess whitespace causes our text-width computation to think the key text is wider than it really is (once rendered within HTML).  The shift layer's keys do not have this excess whitespace.

So, this _compounding_ with the double-scaling effect noted previously caused the drastic shrinkage noted earlier.

... those spaces in the key-text should probably become an issue against that keyboard on our keyboards repo.  While we _could_ try the ol' `String.trim()` approach, this could cause problems for any intentional whitespace; `trim()` will annihilate `&nbsp;` characters, among other things.  It's simple enough to fix within the keyboard's source.

**Comment by jahorton** [comment]
The latest commit should fix all but the missing-font and the excess-headspace issues.

**Comment by mcdurdin** [comment]
> The latest commit should fix all but the missing-font and the excess-headspace issues.

I can confirm that the re-run gave good results.

image

**Review by mcdurdin — APPROVED** [review]

**Comment by jahorton** [comment]
> I can confirm that the re-run gave good results. The mozhi_malayalam keyboard is the only one with serious sizing issues now.

`mozhi-malayalam` is at fault for it, too.  I documented the cause as https://github.com/keymanapp/keyboards/issues/1445.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 14.0.228-beta](https://keyman.com/downloads/releases/beta/14.0.228)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.9-alpha](https://keyman.com/downloads/releases/alpha/15.0.9)

**Pull request merged**